CNV-46630: Connecting VMs to primary UDNs (GA) (#86295)
openshift/openshift-docs: modules/virt-creating-a-primary-udn.adoc

// Module included in the following assemblies:
//
// * virt/vm_networking/virt-connecting-vm-to-primary-udn.adoc

:_mod-docs-content-type: PROCEDURE                                
[id="virt-creating-a-primary-udn_{context}"]                                  
= Creating a primary namespace-scoped user-defined network

You can create an isolated primary network in your project namespace. You must use the OVN-Kubernetes layer 2 topology and enable persistent IP address allocation in the user-defined network (UDN) configuration to ensure VM live migration support.

.Prerequisites
. You have installed the OpenShift CLI (`oc`).

.Procedure
. Create a `UserDefinedNetwork` object to specify the custom network configuration:
+
.Example `UserDefinedNetwork` manifest
[source,yaml]
----
apiVersion: k8s.ovn.org/v1
kind: UserDefinedNetwork
metadata:
  name: udn-l2-net # <1>
  namespace: my-namespace # <2>
spec:
  topology: Layer2 # <3>
  layer2: 
    role: Primary # <4>
    subnets:
      - "10.0.0.0/24"

>       - "2001:db8::/60" 

Reviewer: @ormergi didn't we change this recently ? Or was it just for the c-UDN ?

Author: Updated the YAML

  ipamLifecycle: Persistent # <5>
----
<1> The name of the `UserDefinedNetwork` custom resource. 

> <2> The namespace in which the VM is located. This value cannot be `default` or match any global namespaces that are created by the Cluster Network Operator (CNO).

Reviewer: we need to indicate that this namespace **must** feature the `k8s.ovn.org/primary-user-defined-network`, right @phoracek ?

> <2> The namespace in which the VM is located. This value cannot be `default` or match any global namespaces that are created by the Cluster Network Operator (CNO).

Reviewer 2: ```suggestion
<2> The namespace in which the VM is located. This value should not be `default` or match any global namespaces defied by Cluster Network Operator (CNO).
```

Reviewer: *defined

<3> The topological configuration of the network. The required value is `Layer2`. A `Layer2` topology creates a logical switch that is shared by all nodes.
<4> Specifies if the UDN is primary or secondary. {VirtProductName} only supports the `Primary` role. This means that the UDN acts as the primary network for the VM and all default traffic passes through this network.
<5> Specifies that virtual workloads have consistent IP addresses across reboots and migration. The `spec.layer2.subnets` field is required when `ipamLifecycle: Persistent` is specified.

. Apply the `UserDefinedNetwork` manifest by running the following command:
+
[source,terminal]
----

> $ oc apply -f <filename>.yaml

Reviewer 2: Please see https://github.com/openshift/openshift-docs/pull/86295/files#r1924126560

// Module included in the following assemblies:
//
// * virt/vm_networking/virt-connecting-vm-to-primary-udn.adoc

:_mod-docs-content-type: PROCEDURE                                
[id="virt-creating-a-primary-udn_{context}"]                                  
= Creating a primary namespace-scoped user-defined network

You can create an isolated primary network in your project namespace. You must use the OVN-Kubernetes layer 2 topology and enable persistent IP address allocation in the user-defined network (UDN) configuration to ensure VM live migration support.

> .Prerequisites

Reviewer: @phoracek do we need any sort of clarification / differentiation for a non-admin user ?

Reviewer 3: I would leave it as is, not to overcomplicate the guide. As far as I'm concerned, the whole process of defining a network is for the cluster admin persona.

. You have installed the OpenShift CLI (`oc`).

> . You have created a namespace and applied the `k8s.ovn.org/primary-user-defined-network` label.

Reviewer 4: @maiqueb This is new to me.
Until now, we only needed to create a UDN in a certain namespace, in order to make this namespace a UDN one.
Do you know anything else?

Reviewer: Yes, this is new.

You need to do this upon namespace creation, since the namespace is not mutable. 

Hence, the user, when they create the namespace, must already label the namespace appropriately. 

Afterwards, they'll create the UDN. This was the way we found of preventing the UDN / workload race where the workload would be processed before the UDN, and the workload would be granted a default cluster network attachment instead.

Reviewer 5: Is this also true for C-UDN?

.Procedure
. Create a `UserDefinedNetwork` object to specify the custom network configuration:
+
.Example `UserDefinedNetwork` manifest
[source,yaml]
----
apiVersion: k8s.ovn.org/v1
kind: UserDefinedNetwork
metadata:
  name: udn-l2-net # <1>
  namespace: my-namespace # <2>
spec:
  topology: Layer2 # <3>
  layer2: 
    role: Primary # <4>
    subnets:
      - "10.0.0.0/24"
      - "2001:db8::/60" 

>   ipam

Reviewer 3: this is missing colon

# ...
----
<1> The name of the `UserDefinedNetwork` custom resource. 
<2> The namespace in which the VM is located. This value cannot be `default` or match any global namespaces that are created by the Cluster Network Operator (CNO).

> <3> The topological configuration of the network. The required value is `Layer2`. A `Layer2` topology creates a logical switch that is shared by all nodes.

Reviewer: let's not even mention Layer3. Live migration will **not** work with that.

> <2> The namespace in which the VM is located. The namespace must have the `k8s.ovn.org/primary-user-defined-network` label. The namespace must not be `default`, an `openshift-*` namespace, or match any global namespaces that are defined by the Cluster Network Operator (CNO).

Reviewer 6: Maybe keep consistent with the previous.
```suggestion
<2> The namespace in which the VM is located. The namespace must have the `k8s.ovn.org/primary-user-defined-network` label.  The namespace selector must not point to `default`, an `openshift-*` namespace, or any global namespace that is defined by the Cluster Network Operator (CNO).
```

Author: @mburke5678 This call out describes the actual namespace value while the call out in the other procedure is for a namespace selector that points to namespace values in an array. The two fields are different.

> <1> The name of the `UserDefinedNetwork` custom resource. 
> <2> The namespace in which the VM is located. The namespace must have the `k8s.ovn.org/primary-user-defined-network` label. The namespace must not be `default`, an `openshift-*` namespace, or match any global namespaces that are defined by the Cluster Network Operator (CNO).
> <3> The topological configuration of the network. The required value is `Layer2`. A `Layer2` topology creates a logical switch that is shared by all nodes.
> <4> Specifies if the UDN is primary or secondary. {VirtProductName} only supports the `Primary` role. This means that the UDN acts as the primary network for the VM and all default traffic passes through this network.
> <5> Specifies that virtual workloads have consistent IP addresses across reboots and migration. The `spec.layer2.subnets` field is required when `ipam.lifecycle: Persistent` is specified.

Reviewer 6: Same here
```suggestion
<1> Specifies the name of the `UserDefinedNetwork` custom resource. 
<2> Specifies the namespace in which the VM is located. The namespace must have the `k8s.ovn.org/primary-user-defined-network` label. The namespace must not be `default`, an `openshift-*` namespace, or match any global namespaces that are defined by the Cluster Network Operator (CNO).
<3> Specifies the topological configuration of the network. The required value is `Layer2`. A `Layer2` topology creates a logical switch that is shared by all nodes.
<4> Specifies if the UDN is primary or secondary. {VirtProductName} only supports the `Primary` role. This means that the UDN acts as the primary network for the VM and all default traffic passes through this network.
<5> Specifies that virtual workloads have consistent IP addresses across reboots and migration. The `spec.layer2.subnets` field is required when `ipam.lifecycle: Persistent` is specified.
```
